Khan/wonder-blocks
# Experiment with TS language service plugins

## Summary:
The TypeScript language service supports plugins that can provide additional diagnostics.  This PR adds plugins to run eslint and type coverage.  Although VSCode plugins for eslint and type coverage exist, running them via plugins offers the following benefits:
- easier set up, devs don't have to install extra plugins
- a single point a failure, easier to notice when the service dies because all diagnostics go out together
- provides type coverage to editors/IDEs that don't have a type coverage plugin (JetBrains editors, vim, etc.)

I had to disable "typescript.tsserver.experimental.enableProjectDiagnostics" in .vscode/settings.json.  This option causes all files to be checked by the TS server even if the files aren't open.  As a result, eslint was running on files in node_modules even though those files are ignored.

People can still get project wide TS errors by running 'yarn tsc -w' in a terminal.

Issue: None

## Test plan:
- disable the eslint and type coverage VSCode plugins
- add some lint to a file
- add some uncovered code to file
- see that it's detected by the plugins

<img width="729" alt="Screen Shot 2023-05-11 at 11 11 32 PM" src="https://github.com/Khan/wonder-blocks/assets/1044413/50476b3d-2f13-42fc-88b5-8faca8907c98">

<img width="768" alt="Screen Shot 2023-05-11 at 11 11 22 PM" src="https://github.com/Khan/wonder-blocks/assets/1044413/461bdfa5-7775-45cc-ab98-021f7d9f9a61">

<img width="829" alt="Screen Shot 2023-05-11 at 11 13 15 PM" src="https://github.com/Khan/wonder-blocks/assets/1044413/7e1abaee-333d-45bd-ad1f-cda700a353aa">

## Performance
I used the following code for the type checking (toggling `y`'s type between `string` and `number`) and type coverage (toggling `x`'s type between `any` and `number`) tests:
```
export function foo(x: any, y: string) {
    return x + y;
}

foo(5, 10);
```

For the eslint tests, I added a `console.log()` by pasting it and seeing how long it took for the new lint to be reported.

NOTE: The tools reported diagnostics in response to either document update or document save events.  I've noted this in parentheses for each test.

### VSCode plugins:
- type checking (on update): 3.85s, 4.50s, 3.36s, 4.06s
- type coverage (on save): 3.87s, 3.93s, 3.38s, 4.30s
- eslint (on save): almost instantaneous, but requires a save

### TypeScript plugins:
- type checking (on update): 2.93, 2.90s, 2.78s, 3.17s
- type coverage (on save): 2.30s, 2.91s, 3.38s, 2.76s
- eslint (on update): 2.80s, 3.00s, 2.50s, 3.25s

NOTE: `typescript-eslint-language-service` can be disabled by setting the `TS_ESLINT_SERVICE_DISABLED` and restarting your editor/IDE.
